### PR TITLE
Account abstraction (WIP)

### LINKED ISSUE #866: [SPEC ONLY]: Gather requirements for fund accounts
In order to be ready for Comsic Elevator development we need to agree upon the scope for the engineering team to deliver in this release.

[Notion Board](https://www.notion.so/vegaprotocol/Vega-mainnet-has-fund-accounts-Fund-accounts-provides-semi-passive-LP-following-liquidity-provisio-c01a87eea6e94cba8c54395d692771f0)

Market making and trading in general require a high level of expertise which means that only a small subset of potential users can engage with these activities. Finance solves for this with the existence of funds (of which there are many types) that allow someone with capital (usually after paying a fee/price) deposit that capital with someone else who will trade on their behalf. This has the benefit of allow more and different types of participants in the marketplace and also of increasing the amount of capital available, for example, to provide liquidity. This increased capital and liquidity benefits all participants through lower spread, reduced slippage, and more viable products/markets.

In addition to people/entities doing this "off-chain" with traditional structures and risks, the blockchain enables funds to be created that are non-custodial, not give their operators direct access to funds, and that enforce certain rules and transparency around trading activities. Vega intends to support such a feature to enable on-chain fund creation.

Any MVP of this feature must enable deplyoment of funds for liquidity provision on Vega, including the fund account making and keeping to an LP commitment. This will enable participants in the network to deploy liquidity to markets without trading expertise. 

How do we know when we are done:

- [ ] A fund account spec has been written
- [ ] The specification contains acceptance criteria with IDs that reflect the scope of work required for Cosmic Elevator
- [ ] The engineering team has reviewed the specifications and understand the scope
- [ ] The QA team has reviewed and understands the specification and approves or has added to the set of ACs
- [ ] The spec is merged into the specs repo before engineering start development/test work

### PR DESCRIPTION
**Pull request by barnabee**
Summary of account abstraction feature for discussion and design and spec work.

tl;dr:

1. Add multisig support with:
    - ability to update signer account IDs
    - ability to change signing threshold
    - ability to separate controller account ID (could be the same or a diff. multisig or single user) and signer accounts
    - multisig account ID can be used anywhere (?) a normal pubkey can be
2. Add fund accounts with:
   - separation of "investor", "controller", and "fund manager/trader" roles (account ID, each can be a multisig, at least investors can be other fund accounts, though with a general abstraction all 3 could be)
   - tracking of users' share of fund NAV
   - optional rules for how long it takes to remove funds
   - ability to apply a "2 and 20" style fee (fraction of NAV per period X + fraction of gains net of losses from prior periods per period X) to be sent to any account ID of fund creator's choosing (can be multisig or another fund)
   - ability to restrict "investors" to a limited set of account IDs or leave open to all
   - ability to cap total "invested" or close to new money coming in
   - ability to restrict the fund to a subset of assets/markets/funds
   - ability to restrict the fund to a subset of trader/LP/fund investor roles
   - fund account ID can be used anywhere (?) a normal pubkey can be as long as not restricted by fund rules
   - delay for changing fund rules >= delay for redemption? 
3. Ability to create a "meta" account that allows other accounts do a a rules based subset of activities
    - rules control other accounts ability to act "as" the meta account for any normal tx type
    - can therefore subdivide one "meta" account to diffrent accounts for each type of rule set / responsibility
    - can restrict to specific tx types, markets, assets, etc.

### CODE REVIEW

**Comment by barnabee** [comment]
Thinking about it, we can probably live without item 3 of the tl;dr in the first instance, though I'd need to discuss with some MMs and other community members to understand the specific needs of their use cases better.

File: 0000-account-abstraction.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+- The first type of account will be a standard public/private keypair.
+
+- The other three are described below.
+
+- There will need to be appropriate core and data node APIs for querying accounts and connecting between accounts (for example getting the account ID for a public key).
+
+- Will need a way to wrap to inform Vega of the account to be used.
+  For example, whether I am acting theough my pubkey account, a multisig, fund account, etc.
+	To prevent excessive complexity, layered wrapping of account types could be done when setting an account up but it would probably make sense to generate a single account ID for each potential destination at that point. Note to review and get comfortable with the approach here.
```
Comment (witgaw) [inline 0000-account-abstraction.md:49]: Could you clarify the bit about "generate a single account ID for each potential destination". 
Does it mean account B acting on behalf of account A gets labeled as account C everywhere or something else?

File: 0000-account-abstraction.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  Their share is reduced by the fraction of the asset that is removed.
+
+- A fund account has a manager account ID.
+
+- The manager account can deploy funds through trading, liquidity provision, transfers, or by placing funds in other fund accounts.
+  Actions of the manager account may be delegated or restricted by making it permissioned account.
+	The actions may be performed by a DAO or committee by making it a multisig.
+  The manager account could be a multisig with permissions applied, or some combinations of permissions and several multisigs, etc. to create more complex structures.
+
+- The
```
Comment (witgaw) [inline 0000-account-abstraction.md:212]: How would fee distribution work? Get the total value of general + margin account balances for the fund account and transfer a given percentage from general if available? I suppose it'd be up to the fund account manager to assure enough is available there (so no auto-liquidation or anything like that).

File: 0000-account-abstraction.md (hunk 1 above)
Comment (barnabee) [reply 0000-account-abstraction.md:49]: Yes I think it means something like that. I am a bit unsure about this bit and happy for you to rip any of it up and start again if you can comeup with something better.

File: 0000-account-abstraction.md (hunk 2 above)
Comment (barnabee) [reply 0000-account-abstraction.md:212]: Good question. That sounds like the simplest option.

File: 0000-account-abstraction.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+- The controller can update the weights of each account.
+
+- The controller can change the approval number threshold.
+
+- The controller can change the approcal weight-fraction threshold.
+
+- The controller can change the account-wide timeout.
+
+
+## Account permissions and restrictions
```
Comment (barnabee) [inline 0000-account-abstraction.md:107]: See my comment on the PR @witgaw, I think it'd be good to descope this from this PR and include basic restrictions of scope of use of funds in the fund account spec directly.
